rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,453 @@
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self, msg: MQTTMessage) -> MQTTTransaction {
+        let mut tx = MQTTTransaction::new(msg);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> bool {
+        if input.len() == 0 {
+            return true;
+        }
+
+        self.request_buffer.extend(input);
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.request_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("request: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({} => '{:?}')", current.len(), rem.len(), rem);
+                    current = rem;
+
+                    if let MQTTOperation::CONNECT(ref conn) = msg.op {
+                        self.protocol_version = conn.protocol_version;
+                    }
+                    let tx = self.new_tx(msg);
+                    self.transactions.push(tx);
+                }
+                Err(nom::Err::Incomplete(v)) => {
+                    if let nom::Needed::Size(s) = v {
+                        SCLogDebug!("incomplete request: needed {}", s);
+                    }
+                    self.request_buffer.extend_from_slice(current);
+                    break;
+                }
+                Err(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> bool {
+        if input.len() == 0 {
+            return true;
+        }
+
+        self.response_buffer.extend(input);
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.response_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({})", current.len(), rem.len());
+                    current = rem;
+                    let tx = self.new_tx(msg);
```

Comment:
it looks like there is no attempt to match requests and responses in a single transaction?

Reply:
Correct, the current implementation treats each single message as its own transaction. Linking request/responses when using higher QoS levels is something was thinking of (see my previous PR) but I'm not sure whether this adds a lot of value atm.

Reply:
In general, Suricata expects transactions to be bidirectional (to be a transaction instead of merely a record). There are some parsers that don't do this, and they'll have issues pretty soon with some upcoming changes around conditional logging. So a bidirectional transaction should normally be the default.

Reply:
Okay. As I said, this will only be relevant for request QoS levels that are higher than 0 ("fire and forget") which do not have a response. But I will look into adding this. My approach would be to have multiple sub-objects in the log item, one for each message in the transaction, e.g.:
```json
"mqtt": {
  "retain": false,
  "dup": false,
  "publish": {
    "qos": 1,
    "topic": "topic",
    "message_id": 1,
    "message": "baa baa sheep"
  },
  "puback": {
    "qos": 0,
    "message_id": 1,
    "reason_code": 16
  }
}
```

Reply:
This might take a while as it will also impact almost all MQTT detection code, just FYI.

Reply:
Phew, done. In next MR.